p2p/channel.go

```diff
@@ -80,125 +85,251 @@ type stream struct {
 	resCh chan *transportMsg
 }
 
+type peer struct {
+	sync.RWMutex
+	publicKey  PublicKey
+	remoteAddr *net.UDPAddr
+}
+
+func (p *peer) addr() *net.UDPAddr {
+	p.RLock()
+	defer p.RUnlock()
+
+	return p.remoteAddr
+}
+
+func (p *peer) updateAddr(addr *net.UDPAddr) {
+	p.Lock()
+	defer p.Unlock()
+
+	p.remoteAddr = addr
+}
+
 // transport wraps network primitives for sending and receiving packets.
 type transport struct {
-	*textproto.Writer
-	*textproto.Reader
+	// textReader is used to read p2p text protocol data.
+	textReader *textproto.Reader
+
+	// textWriter is used to marshal messages to underlying text protocol.
+	textWriter *textproto.Writer
+
+	// session is KCP session which wraps UDP connection and adds reliability and ordered messages support.
 	session *kcp.UDPSession
-	conn    *net.UDPConn
+
+	// remoteConn is initial conn which should be created from NAT hole punching or manually. It contains
+	// initial local and remote peer addresses.
+	remoteConn *net.UDPConn
+
+	// proxyConn is used for KCP session as a remote. Since KCP doesn't expose it's data read loop
+	// this is needed to detect remote peer address changes as we can simply use conn.ReadFromUDP and
+	// get updated peer address.
+	proxyConn *net.UDPConn
 }
 
 // channel implements Channel interface.
 type channel struct {
 	mu   sync.RWMutex
 	once sync.Once
 
-	tr               *transport
-	serviceConn      *net.UDPConn
-	topicHandlers    map[string]HandlerFunc
-	streams          map[uint64]*stream
-	nextStreamID     uint64
-	privateKey       PrivateKey
-	peerPubKey       PublicKey
-	peerAddr         *net.UDPAddr
-	localAddr        *net.UDPAddr
-	blockCrypt       kcp.BlockCrypt
-	stop             chan struct{}
-	sendQueue        chan *transportMsg
+	// tr is transport containing network related connections for p2p to work.
```

Comment:
These obvious comments for private fields looks excess.

Reply:
imho its good. We don't need just formal documentation. Its sometimes better to have more detailed explanations for more complex code parts.